addons/pos_coupon/static/src/js/coupon.js

```diff
@@ -1017,20 +1017,30 @@ odoo.define('pos_coupon.pos', function (require) {
                 .join(',');
         },
         _createDiscountRewards: function (program, coupon_id, amountsToDiscount) {
-            const discountRewards = Object.entries(amountsToDiscount).map(([tax_keys, amount]) => {
+            const rewards = [];
+            let totalDiscount = 0;
+            for (let [tax_keys, amount] of Object.entries(amountsToDiscount)) {
                 let discountAmount = (amount * program.discount_percentage) / 100.0;
+                if (totalDiscount >= program.discount_max_amount && program.discount_max_amount > 0) {
+                    break;
```

Comment:
This is wrong because you are skipping discount with specific tax. Let's say you have Product A and product B with different taxes. If a discount is applied, let's say -50%, there should be two discount lines with different values and taxes. So if the price of A is 10€ and B is 20€, you will have a discount line of -5€ with tax A and another discount line of -10€ with tax B. In your case, you will only have one discount line if the discount max amount is reached.

So what you need to do here is to properly divide the total discount between the discount lines considering the weight they hold in the total amount of the order (in the example, product B is more expensive than A so the discount line with its tax should be higher than A)